Issue 21810/improve validation methods

glowe opened the pull request:
<!-- Please describe your changes on the following line: -->
This is a start at addressing #21810. I'm putting these changes out early to get some feedback on the following items:

1. I added unit tests for the validation methods mentioned in #21810, because I couldn't tell whether any of the existing WPT tests covered them. Are these tests worthwhile? Are any of them unnecessary?
2. I changed the implementation for `is_valid_floating_point_number_string` so that it passed the tests. The previous version of the function wasn't restrictive enough (it allowed certain whitespace characters before the number string).
3. I changed the catch-all condition in `htmlinputelement.rs` to account for the remaining input types that don't have a value sanitization algorithm. This last change seems good to me since we won't be able to add a new input type without adding it to the case and checking the spec for an algorithm.
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #21810

<!-- Either: -->
- [x] There are tests for these changes 

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @asajeffrey (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmlinputelement.rs, components/script/dom/bindings/str.rs
 * @kichjang: components/script/dom/htmlinputelement.rs, components/script/dom/bindings/str.rs

jdm reviewed (approved):
I'm in favour of unit tests for small, self-contained routines like this.

jdm commented on components/script/dom/bindings/str.rs:
```diff
@@ -322,10 +322,79 @@ impl DOMString {
+    /// MM both must be two digits, where 1 <= MM <= 12
     /// https://html.spec.whatwg.org/multipage/#valid-month-string
     pub fn is_valid_month_string(&self) -> bool {
-        parse_month_string(&self.0).is_ok()
```
Whoops, I missed reading this commit. While in general I'm in favour of using state machines for clarity of purpose and avoiding invalid states, I find this one to be difficult to follow. I'm not sure if it's the names of the states, or the repetition between them, or something else, but my eyes keep glazing over when I try to follow the logic. Given how straightforward the specification text is, I'm not exactly sure what to recommend.

jdm replied:
My reading of the specification is that the parsing algorithm for month strings is the same as the validation algorithm for month strings, so I don't see a good reason to reimplement this particular algorithm. Presumably there are other input types where they diverge?

glowe replied:
> My reading of the specification is that the parsing algorithm for month strings is the same as the validation algorithm for month strings, so I don't see a good reason to reimplement this particular algorithm. Presumably there are other input types where they diverge?

Actually, the tests passed for all of the original implementations, so I judge them to be equivalent to the specifications. The original issue #21810 recommended reimplementing all of these functions, but I think that recommendation was based on the premise that the functions *might* be incorrect. 

To move things forward, how about dropping 8a96009 from this PR?

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 8abbae1 has been approved by `jdm`

<!-- @bors-servo r=jdm 8abbae16c131c95b424add550bdaf2700c8e0403 -->

bors-servo commented:
:hourglass: Testing commit 8abbae16c131c95b424add550bdaf2700c8e0403 with merge da3f0212000ce4f83a8dcdd4dc19feb78eef8ada...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/PS2ccpM_RpO0M9rhM00ZcA)

jdm commented:
Hmm, looks like we now fail a test that we used to pass:
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "assert_equals: expected \"\" but got \"2e308\"", 
    "stack": "@http://web-platform.test:8000/html/semantics/forms/the-input-element/number.html:50:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1931:25\ntest@http://web-platform.test:8000/resources/testharness.js:544:30\n@http://web-platform.test:8000/html/semantics/forms/the-input-element/number.html:43:5\n", 
    "subtest": "value >= Number.MAX_VALUE", 
    "test": "/html/semantics/forms/the-input-element/number.html", 
    "line": 60091, 
    "action": "test_result", 
    "expected": "PASS"
}
```
This can be reproduced with `./mach test-wpt tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/number.html`.

glowe commented:
> This can be reproduced with ./mach test-wpt tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/number.html.

Thanks for identifying the failing test for me.

We are now failing on a bounds check: `2e308` exceeds `Number.MAX_VALUE` (`1.7976931348623157e+308`). I could be reading this wrong, but the specification for a valid floating point number (https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-floating-point-number) doesn't mention a bounds. However the specification for parsing (https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-floating-point-number-values) does mention it in that the converted value must be a `IEEE 754 double-precision floating-point value`.

_Interestingly, I tried this specific test on Firefox 70.0.1 and it failed (http://wpt.live/html/semantics/forms/the-input-element/number.html). However, this test passes in the latest Chrome and Safari browsers._

Here's what I propose we do next (let me know what you think):

1.) I'll update the code to handle the bounds check, probably by delegating to the parsing code after verifying that there's no leading whitespace.
2.) I'm now wondering about the value of the unit tests I added. I didn't notice these web platform tests before and it seems like the unit tests are duplicative and potentially misleading. I'm leaning towards dropping the unit tests for this PR.
3.) The web platform tests for number inputs are incomplete for the leading whitespace check:

https://github.com/servo/servo/blob/b62d1d4f7a70292b55547f6ea9f59e19525522ec/tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/number.html#L38

The specification mentions that the whitespace characters can be `TAB`, `LF`, `FF`, `CR`, or `SPACE` (https://infra.spec.whatwg.org/#ascii-whitespace). Our previous implementation only handled `SPACE`s. Would it be worth me filing an issue (and potentially a PR) to add tests for the other characters?

Commit:
Remove catch-all case for input sanitization

Replaced catch-all with explicit case for inputs that do not have
a value sanitization algorithm. This should prevent us from
forgetting to implement a sanitization for an input, since they
must all be accounted for in the match expression.

Commit:
Add DOMString floating point string test and fix

Fixes an issue where DOMString::is_valid_floating_point_number_string
was returning true for strings that began with whitespace characters-
TAB, LF, FF, or CR. Also added a unit test to cover this since the
corresponding web-platform-tests are incomplete.

jdm commented:
Those choices seem reasonable! You can also include the WPT changes in this PR, because we automatically sync them upstream when they merge in our vendored copy.

jdm commented:
And FYI https://wpt.fyi/results/html/semantics/forms/the-input-element/number.html?label=master&label=experimental&aligned&q=number.html backs up your results with respect to Firefox's test results.

Commit:
Enhance wpt for number input

Add cases for all leading ASCII whitespace characters.

Commit:
Remove redundant domstring unit test

This test is no longer necessary since the cases tested are also tested
by the number input web platform test.

glowe commented:
> Those choices seem reasonable! You can also include the WPT changes in this PR, because we automatically sync them upstream when they merge in our vendored copy.

Thanks! I did not know that. I've enhanced the number input WPT to include those additional cases and deleted the unit test since it's redundant coverage.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/20575.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit b9ec6f9 has been approved by `jdm`

<!-- @bors-servo r=jdm b9ec6f99cd76ed430d5916b38190e8ee7be70516 -->

bors-servo commented:
:hourglass: Testing commit b9ec6f99cd76ed430d5916b38190e8ee7be70516 with merge 7ba88e823744c0d1a37fdd1a00f3571d48c698bb...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/XmC49avFRBiSk0_8TJuiTg)
Approved by: jdm
Pushing 7ba88e823744c0d1a37fdd1a00f3571d48c698bb to master...
